viennacl/ocl/mempool/mempool.hpp

```diff
@@ -0,0 +1,357 @@
+// Abstract memory pool implementation
+//
+// Copyright (C) 2009-17 Andreas Kloeckner
+//               2018-19 Kaushik Kulkarni
+//
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+
+#ifndef VIENNACL_OCL_MEMPOOL_HPP_
+#define VIENNACL_OCL_MEMPOOL_HPP_
+
+#ifdef __APPLE__
+#include <OpenCL/cl.h>
+#else
+#include <CL/cl.h>
+#endif
+
+#include <memory>
+#include <vector>
+#include <map>
+#include <iostream>
+#include <cassert>
+
+#include "viennacl/ocl/mempool/bitlog.hpp"
+#include "viennacl/ocl/mempool/mempool_utils.hpp"
+#include "viennacl/ocl/error.hpp"
+
+namespace viennacl
+{
+namespace ocl
+{
+
+  template<class Allocator>
+  class memory_pool : mempool::noncopyable
+  {
+    public:
+      typedef size_t size_type;
+
+    private:
+      typedef uint32_t bin_nr_t;
+      typedef std::vector<cl_mem> bin_t;
```

Comment:
The original PyOpenCL memory pool was not backend-specific--in fact, it was used identically (same header file!) in PyCUDA. How come `cl_mem` can't be obtained from the allocator type as in the original pool?